dev/com.ibm.ws.install/src/com/ibm/ws/install/maven/MavenDirectoryInstall.java

```diff
@@ -0,0 +1,276 @@
+
+		map = new InstallKernelMap();
+		initializeMap(featuresToInstall);
+		getLibertyVersions();
+
+	}
+
+	public MavenDirectoryInstall(Collection<String> featuresToInstall, File fromDir, String toExtension)
+			throws IOException, InstallException {
+		this(featuresToInstall, fromDir);
+		this.toExtension = toExtension;
+
+	}
+
+	/**
+	 * Initialize the Install kernel map.
+	 * 
+	 * @param featuresToInstall
+	 * @throws IOException
+	 */
+	private void initializeMap(Collection<String> featuresToInstall) throws IOException {
+		map.put("runtime.install.dir", Utils.getInstallDir());
+		map.put("target.user.directory", new File(Utils.getInstallDir(), "usr/tmp"));
+		map.put("install.local.esa", true);
+		map.put("single.json.file", (getSingleJsonPaths(fromDir)));
+		map.put("features.to.resolve", new ArrayList<>(featuresToInstall));
+		map.put("license.accept", true); // TODO: discuss later
+		map.get("install.kernel.init.code");
+
+	}
+
+	/**
+	 * Get the open liberty runtime version.
+	 * 
+	 * @throws IOException
+	 * @throws InstallException
+	 * 
+	 */
+	private void getLibertyVersions() throws IOException, InstallException {
+		File dir = new File(Utils.getInstallDir(), "lib/versions");
+		File[] propertiesFiles = dir.listFiles(new FilenameFilter() {
+			@Override
+			public boolean accept(File dir, String name) {
+				return name.endsWith(".properties");
+			}
+		});
+
+		if (propertiesFiles == null) {
+			throw new IOException("Could not find properties files.");
+		}
+
+		for (File propertiesFile : propertiesFiles) {
+			Properties properties = new Properties();
+			try (InputStream input = new FileInputStream(propertiesFile)) {
+				properties.load(input);
+				String productId = properties.getProperty("com.ibm.websphere.productId");
+				String productVersion = properties.getProperty("com.ibm.websphere.productVersion");
+				
+				if (productId.equals(OPEN_LIBERTY_PRODUCT_ID)) {
+					openLibertyVersion = productVersion;
+				}
+
+			} catch (IOException e) {
+				throw new IOException("Could not read the properties file " + propertiesFile.getAbsolutePath());
+			}
+		}
+		if (openLibertyVersion == null) {
+			// openliberty.properties file is missing
+			throw new InstallException("Could not determine the open liberty runtime version.");
+		} else {
+			logger.log(Level.FINE, "The Open Liberty runtime version is " + openLibertyVersion);
+		}
+
+	}
+
+	/**
+	 * Resolves and installs the features
+	 *
+	 * @throws InstallException
+	 * @throws IOException
+	 */
+	@SuppressWarnings("unchecked")
+	public void installFeatures() throws InstallException, IOException {
+		Collection<String> resolvedFeatures = (Collection<String>) map.get("action.result");
+		logger.log(Level.FINE, "Resolved features: " + resolvedFeatures);
+		if (resolvedFeatures == null) {
+			throw new InstallException((String) map.get("action.error.message")); // TODO: appropriate message
```

Comment:
Replace with existing or new message

Reply:
The install kernel and install kernel map handle the exception already